src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs

```diff
@@ -257,8 +258,331 @@ private struct IntPtrAndHandle
+            ARRAY_UBYTE = 11,
+            ARRAY_UBYTE_C = 12,
+            ARRAY_SHORT = 13,
+            ARRAY_USHORT = 14,
+            ARRAY_INT = 15,
+            ARRAY_UINT = 16,
+            ARRAY_FLOAT = 17,
+            ARRAY_DOUBLE = 18,
+            FP32 = 24,
+            UINT32 = 25,
+            INT64 = 26,
+            UINT64 = 27,
+            CHAR = 28,
+            STRING_INTERNED = 29,
+            VOID = 30,
+            ENUM64 = 31
+        }
+
+        // see src/mono/wasm/driver.c MARSHAL_ERROR_xxx
+        public enum MarshalError : int {
+            BUFFER_TOO_SMALL = 512,
+            NULL_CLASS_POINTER = 513,
+            NULL_TYPE_POINTER = 514,
+            UNSUPPORTED_TYPE = 515,
+            FIRST = BUFFER_TOO_SMALL
+        }
+
+        public static string MakeMarshalTypeRecord (Type type) {
+            return $"{{ \"marshalType\": {(int)GetMarshalTypeFromType(type)}, \"typePtr\": {type.TypeHandle.Value} }}";
+        }
+
+        public static unsafe string? MakeMarshalSignatureInfo (IntPtr typePtr, IntPtr methodPtr) {
+            if (methodPtr == IntPtr.Zero)
+                return null;
+
+            IntPtrAndHandle tmp = default(IntPtrAndHandle);
+            tmp.ptr = methodPtr;
+            var methodHandle = tmp.handle;
+            tmp.ptr = typePtr;
+            var typeHandle = tmp.typeHandle;
+
+            var type = (typePtr != IntPtr.Zero)
+                ? Type.GetTypeFromHandle(typeHandle)
+                : null;
+
+            MethodBase? mb = (typePtr != IntPtr.Zero)
+                ? MethodBase.GetMethodFromHandle(methodHandle, typeHandle)
+                : MethodBase.GetMethodFromHandle(methodHandle);
+            if (mb == null)
+                return null;
+
+            var result = "{ " +
+                $"\"result\": {MakeMarshalTypeRecord((mb as MethodInfo)?.ReturnType ?? typeof(void))}, " +
+                $"\"typePtr\": {typePtr}, \"methodPtr\": {methodPtr}, " +
+                "\"parameters\": [";
+
+            int i = 0;
+            foreach (var p in mb.GetParameters()) {
+                if (i > 0)
+                    result += ", ";
+                result += MakeMarshalTypeRecord(p.ParameterType);
+                i++;
+            }
+
+            result += "] }";
+
+            return result;
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
+            Justification = "Trimming doesn't affect types eligible for marshalling. Different exception for invalid inputs doesn't matter.")]
+        private static unsafe string GetAndEscapeStringProperty (Type type, string name) {
+            var info = type.GetProperty(
+                name, BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic
+            );
+
+            var value = info?.GetValue(null) as string;
+            if (value == null)
+                return "null";
+
+            var result = "\"";
```

Comment:
What escaping rules are we trying to mimic here?

Reply:
It's parsed using JSON.parse on the other side of the managed/js boundary